src/storage/lock_manager/lock_waiting_queue.rs

```diff
@@ -0,0 +1,950 @@
+    pub lock_wait_token: LockWaitToken,
+    pub req_states: Option<Arc<LockWaitContextSharedState>>,
+    pub current_legacy_wake_up_cnt: Option<usize>,
+    pub key_cb: Option<SyncWrapper<PessimisticLockKeyCallback>>,
+}
+
+impl PartialEq<Self> for LockWaitEntry {
+    fn eq(&self, other: &Self) -> bool {
+        self.parameters.start_ts == other.parameters.start_ts
+    }
+}
+
+impl Eq for LockWaitEntry {}
+
+impl PartialOrd<Self> for LockWaitEntry {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        // Reverse it since the std BinaryHeap is max heap and we want to pop the
+        // minimal.
+        other
+            .parameters
+            .start_ts
+            .partial_cmp(&self.parameters.start_ts)
+    }
+}
+
+impl Ord for LockWaitEntry {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        // Reverse it since the std BinaryHeap is max heap and we want to pop the
+        // minimal.
+        other.parameters.start_ts.cmp(&self.parameters.start_ts)
+    }
+}
+
+pub struct KeyLockWaitState {
+    #[allow(dead_code)]
+    current_lock: kvrpcpb::LockInfo,
+
+    /// The counter of wake up events of legacy pessimistic lock requests
+    /// (`allow_lock_with_conflict == false`). When an lock wait entry is
+    /// pushed to the queue, it records the current counter. The purpose
+    /// is to mark the entries that needs to be woken up after delaying.
+    ///
+    /// Here is an example showing how it works (note that requests in
+    /// the example are all in legacy mode):
+    ///
+    /// Let's denote a lock-wait entry by `(start_ts,
+    /// current_legacy_wake_up_cnt)`. Consider there are three requests with
+    /// start_ts 20, 30, 40 respectively, and they are pushed to the
+    /// queue when the `KeyLockWaitState::legacy_wake_up_cnt` is 0. Then the
+    /// `KeyLockWaitState` is:
+    ///
+    /// ```text
+    /// legacy_wake_up_cnt: 0, queue: [(20, 0), (30, 0), (40, 0)]
+    /// ```
+    ///
+    /// Then the lock on the key is released. We pops the first entry in the
+    /// queue to wake it up, and then schedule a call to
+    /// [`LockWaitQueues::delayed_notify_all`] after delaying for
+    /// `wake_up_delay_duration`. The current state becomes:
+    ///
+    /// ```text
+    /// legacy_wake_up_cnt: 1, queue: [(30, 0), (40, 0)]
+    /// ````
+    ///
+    /// Here, if some other request arrives, one of them may successfully
+    /// acquire the lock and others are pushed to the queue. the state
+    /// becomes:
+    ///
+    /// ```text
+    /// legacy_wake_up_cnt: 1, queue: [(30, 0), (40, 0), (50, 1), (60, 1)]
+    /// ```
+    ///
+    /// Then `wake_up_delay_duration` is elapsed since the previous waking up.
+    /// Here, we expect that the lock wait entries 30 and 40 can be woken
+    /// up, since they exists when the previous waking up occurs. But we
+    /// don't want to wake up later-arrived entries (50 and 60) since it
```

Comment:
Since 50 and 60 is pushed to the queue, they must have seen some lock on the key. They are added *after* the previous waking up, therefore there must be someone else who acquired the lock after it, otherwise one of 50 and 60 should have successfully acquired the lock instead of started waiting.

Reply:
Gotcha